extras/FreeType/contrib/ttf2pk/ttf2tfm.1
.\" man page for ttf2tfm
.
.TH TTF2TFM 1 25-Jul-1998 "FreeType version 1.2"
.SH NAME
ttf2tfm \- build TeX metric files from a TrueType font
.SH SYNOPSIS
.na
.nh
.B ttf2tfm
.in +\n(.ku
.sp -1
.IR ttffile [ .ttf | .ttc ]
[\c
.BI -c \ \%caps-height\c
]
[\c
.BI -e \ \%extension-factor\c
]
[\c
.BI -E \ \%encoding-id\^\c
]
[\c
.BI -f \ \%font-index\c
]
[\c
.B -n\c
]
[\c
.B -N\c
]
[\c
.B -O\c
]
[\c
.B -p\ \c
.IR \%inencfile [ .enc ]\c
]
[\c
.BI -P \ \%platform-id\^\c
]
[\c
.B -q\c
]
[\c
.BI -r \ \%old-glyphname\ \%new-glyphname\c
]
[\c
.B -R\ \c
.IR \%replacement-file [ .rpl ]\c
]
[\c
.BI -s \ \%slant-factor\c
]
[\c
.B -t\ \c
.IR \%outencfile [ .enc ]\c
]
[\c
.B -T\ \c
.IR \%inoutencfile [ .enc ]\c
]
[\c
.B -u\c
]
[\c
.B -v\ \c
.IR \%vplfile [ .vpl ]\c
]
[\c
.B -V\ \c
.IR \%scvplfile [ .vpl ]\c
]
[\c
.IR \%tfmfile [ .tfm ]\c
]
.br
.in
.B "ttf2tfm --version"
|
.B --help
.ad
.hy
.
.
.
.\" ====
.\" ==== macro definitions
.\" ====
.
.\" here we define \TeX for troff and nroff
.if t .ds TX \fRT\\h'-0.1667m'\\v'0.20v'E\\v'-0.20v'\\h'-0.125m'X\fP
.if n .ds TX TeX
.
.\" and here the same for \LaTeX
.if t \{\
.ie '\*(.T'dvi' \
.ds LX \fRL\h'-0.36m'\v'-0.15v'\s-3A\s0\h'-0.15m'\v'0.15v'\fP\*(TX
.el .ds LX \fRL\h'-0.36m'\v'-0.22v'\s-2A\s0\h'-0.15m'\v'0.22v'\fP\*(TX
.\}
.if n .ds LX LaTeX
.
.\" \LaTeXe
.\" note that we need \vareps for TeX instead of \eps which can only be
.\" accessed with the \N escape sequence (in the Math Italic font)
.if t \{\
.ie '\*(.T'dvi' .ds LE \*(LX\h'0.15m'2\v'0.20v'\f(MI\N'34'\fP\v'-0.20v'
.el .ds LE \*(LX\h'0.15m'2\v'0.20v'\(*e\v'-0.20v'
.\}
.if n .ds LE LaTeX\ 2e
.
.\" a typewriter font
.if t \{\
.de C
\fC\\$1\fP\\$2\fC\\$3\fP\\$4
..
.\}
.if n \{\
.de C
\\$1\\$2\\$3\\$4
..
.\}
.
.\" ====
.\" ==== end of macro definitions
.\" ====
.
.
.
.SH DESCRIPTION
This program extracts the metric and kerning information of a TrueType
font and converts it into metric files usable by \*(TX
(quite similar to
.B afm2tfm
which is part of the
.B dvips
package).
Please consult the documentation of
.B afm2tfm
in the
.B dvips
info file for more details on the various parameters (especially
encoding files).
Here we concentrate on the differences between
.B afm2tfm
and
.BR ttf2tfm .
.PP
Since a TrueType font often contains more than 256\ glyphs, some means
are necessary to map a subset of the TrueType glyphs into a \*(TX
font.
To do this, two mapping tables are needed: the first (called `input' or
`raw' encoding) maps from the TrueType font to a raw \*(TX
font (this mapping table is used both by
.B ttf2tfm
and
.BR ttf2pk ),
and the second (called `output' or `virtual' encoding) maps from the raw
\*(TX
font to another (virtual) \*(TX
font providing all kerning and ligature information needed by \*(TX.
.PP
This two stage mapping has the advantage that one raw font can be
accessed with various \*(TX
encodings (e.g.\ T1 and OT1) via the virtual font mechanism, and just
one
.C PK
file is necessary.
.PP
For CJKV (Chinese/Japanese/Korean/old Vietnamese) fonts, a different
mechanism is provided (see
.B "SUBFONT DEFINITION FILES"
below).
.
.
.SH PARAMETERS
Most of the command line switch names are the same as in
.B afm2tfm
for convenience.
One or more space characters between an option and its parameter is
mandatory; options can't be concatenated.
.TP
.BI -c \ caps-height
The height of small caps made with the
.B -V
switch.
Default value of this real number is 0.8.
Will be ignored in subfont mode.
.TP
.BI -e \ extension-factor
The extension factor to stretch the characters horizontally.
Default value of this real number is 1.0; if less than 1.0, you get a
condensed font.
.TP
.BI -E \ encoding-id
The TrueType encoding ID.
Default value of this integer is\ 1.
Will be ignored if
.B -N
is used.
.TP
.BI -f \ font-index
The font index in a TrueType Collection.
Default is the first font (index 0).
Will be ignored for ordinary TrueType fonts.
[TrueType collections are usually found in some CJK fonts; e.g.\ the first
font index specifies glyphs and metrics for horizontal writing, and the
second font index does the same for vertical writing.]
.TP
.B -n
Use PS names of the TrueType font.
Will be ignored in subfont mode.
Only glyphs with a valid entry in the selected cmap are used.
.TP
.B -N
Use only PS names of the TrueType font.
No cmap is used, thus the switches
.B -E
and
.B -P
have no effect, causing a warning message.
Will be ignored in subfont mode.
.TP
.B -O
Use octal values for all character codes in the
.C VPL
file rather than names; this is useful for symbol or CJK fonts where
character names such as `A' are meaningless.
.TP
.BI -p \ inencfile
The input encoding file name for the TTF\(->raw\ \*(TX
mapping.
This parameter has to be specified in
.C \%ttfonts.map
for successive
.B ttf2pk
calls.
Will be ignored in subfont mode.
.TP
.BI -P \ platform-id
The TrueType platform ID.
Default value of this integer is\ 1.
Will be ignored if
.B -N
is used.
.TP
.B -q
Make
.B ttf2tfm
quiet.
It suppresses any informational output except warning and error
messages.
For CJK fonts, the output can get quite large if you don't specify
this switch.
.TP
.BI -r \ old-glyphname\ new-glyphname
Replaces
.I \%old-glyphname
with
.IR \%new-glyphname .
This switch is useful if you want to give an unnamed glyph (i.e., a glyph
which can be represented with `.gXXX' or `.cXXX' only) a name or if you want
to rename an already existing glyph name.
You can't use the `.gXXX' or `.cXXX' glyph name constructs for
.IR \%new-glyphname ;
multiple occurrences of
.B -r
are possible.
If in subfont mode or if no encoding file is specified, this switch is
ignored.
.TP
.BI -R \ replacement-file
Use this switch if you have many replacement pairs; they can be  collected
in a file which should have `\c
.C \&.rpl '
as extension.
The syntax used in such replacement files is simple: Each non-empty
line must contain a pair `\c
.IR "\%old-glyphname \%new-glyphname" '
separated by whitespace (without the quotation marks).
The percent sign starts a line comment; you can continue a line on the
next line with a backslash as the last character.
If in subfont mode or if no encoding file is specified, this switch is
ignored.
.TP
.BI -s \ slant-factor
The oblique factor to slant the font, usually much smaller than\ 1.
Default of this real number is\ 0.0; if the value is larger than zero,
the characters slope to the right, otherwise to the left.
.TP
.BI -t \ outencfile
The output encoding file name for the virtual font(s).
Only characters in the raw \*(TX
font are used.
Will be ignored in subfont mode.
.TP
.BI -T \ inoutencfile
This is equivalent to
.RB ` -p
.I inoutencfile
.B -t
.IR inoutencfile '.
Will be ignored in subfont mode.
.TP
.B -u
Use only those characters specified in the output encoding, and no
others.
By default,
.B ttf2tfm
tries to include all characters in the virtual font, even those not
present in the encoding for the virtual font (it puts them into
otherwise-unused positions, rather arbitrarily).
Will be ignored in subfont mode.
.TP
.BI -v \ vplfile
Output a
.C VPL
file in addition to the
.C TFM
file.
If no output encoding file is specified,
.B ttf2tfm
uses a default font encoding (cmtt10).
Will be ignored in subfont mode.
.TP
.BI -V \ scvplfile
Same as
.BR -v ,
but the virtual font generated is a pseudo small caps font obtained by
scaling uppercase letters by\ 0.8 (resp. the value specified with
.BR -c )
to typeset lowercase.
This font handles accented letters and retains proper kerning.
Will be ignored in subfont mode.
.TP
.B --version
Shows the current version of
.B ttf2tfm
and the used file search library (e.g.
.BR kpathsea ).
.TP
.B --help
Shows usage information.
.PP
If no
.C TFM
file name is given, the name of the TTF file is used, including the
full path and replacing the extension with `\c
.C \&.tfm '.
.
.
.SH CMAPS
Contrary to Type\ 1 PostScript fonts (but similar to the new CID
PostScript font format), most TrueType fonts have more than one native
mapping table, also called `cmap', which maps the (internal) TTF glyph
indices to the (external) TTF character codes.
Common examples are a mapping table to Unicode encoded character
positions and the standard Macintosh mapping.

To specify this TrueType mapping table, use the options
.B -P
and
.BR -E .
With
.B -P
you specify the platform ID; defined values are:
.PP
.in +4m
.ta 3iC
.I "platform	platform ID (pid)"
.sp
.ta 3iR
Apple Unicode	0
.br
Macintosh	1
.br
ISO	2
.br
Microsoft	3
.PP
The encoding ID depends on the platform.
For pid=0, we ignore the
.B -E
parameter (setting it to zero) since the mapping table is always
Unicode version\ 2.0.
For pid=1, the following table lists the defined values:
.PP
.in +4m
.ta 3iC
.ti -2m
platform ID = 1
.sp
.I "script	encoding ID (eid)"
.sp
.ta 3iR
Roman	0
.br
Japanese	1
.br
Chinese	2
.br
Korean	3
.br
Arabic	4
.br
Hebrew	5
.br
Greek	6
.br
Russian	7
.br
Roman Symbol	8
.br
Devanagari	9
.br
Gurmukhi	10
.br
Gujarati	11
.br
Oriya	12
.br
Bengali	13
.br
Tamil	14
.br
Telugu	15
.br
Kannada	16
.br
Malayalam	17
.br
Sinhalese	18
.br
Burmese	19
.br
Khmer	20
.br
Thai	21
.br
Laotian	22
.br
Georgian	23
.br
Armenian	24
.br
Maldivian	25
.br
Tibetan	26
.br
Mongolian	27
.br
Geez	28
.br
Slavic	29
.br
Vietnamese	30
.br
Sindhi	31
.br
Uninterpreted	32
.PP
Here are the ISO encoding IDs:
.PP
.in +4m
.ta 3iC
.ti -2m
platform ID = 2
.sp
.I "encoding	encoding ID (eid)"
.sp
.ta 3iR
ASCII	0
.br
ISO 10646	1
.br
ISO 8859-1	2
.PP
And finally, the Microsoft encoding IDs:
.PP
.in +4m
.ta 3iC
.ti -2m
platform ID = 3
.sp
.I "encoding	encoding ID (eid)"
.sp
.ta 3iR
Symbol	0
.br
Unicode 2.0	1
.br
Shift JIS	2
.br
GB 2312 (1980)	3
.br
Big 5	4
.br
KSC 5601 (Wansung)	5
.br
KSC 5601 (Johab)	6
.PP
The program will abort if you specify an invalid platform/encoding ID
pair.
Please note that most fonts have at most two or three cmaps, usually
corresponding to the pid/eid pairs (1,0), (3,0), or (3,1) in case of
Latin based fonts.
Valid Microsoft fonts should have a (3,1) mapping table, but some
fonts exist (mostly Asian fonts) which have a (3,1) cmap not encoded
in Unicode.
The reason for this strange behavior is the fact that some MS Windows
versions will reject fonts having a non-Unicode cmap (since all
non-Unicode Microsoft encoding IDs are for Asian specific MS Windows
versions).
.PP
The
.B -P
and
.B -E
options of
.B ttf2tfm
must be equally specified for
.BR ttf2pk ;
the corresponding parameters in
.C \%ttfonts.map
are `Pid' and `Eid', respectively.
.PP
The default pid/eid pair is (3,1).
.PP
Similarly, an
.B -f
option must be specified as `Fontindex' parameter in
.C \%ttfonts.map .
.PP
If you use the
.B -N
switch, all cmaps are ignored, using only the PostScript names in the
TrueType font.
The corresponding option in
.C \%ttfonts.map
is `PS=Only'.
If you use the
.B -n
switch, the default glyph names built into
.B ttf2tfm
are replaced with the PS glyph names found in the font.
In many cases this is not what you want because the glyph names in the
font are often incorrect or non-standard.
The corresponding option in
.C \%ttfonts.map
is `PS=Yes'.
.PP
Single replacement glyphs specified with `\c
.C -r '
must be given directly as
.IR \%old-glyphname = \%new-glyphname ;
`\c
.C -R '
is equivalent to the `Replacement' option.
.
.
.SH INPUT AND OUTPUT ENCODINGS
You must specify the encoding vectors from the TrueType font to the
raw \*(TX
font and from the raw \*(TX
font to the virtual \*(TX
font exactly as with
.BR afm2tfm ,

extras/FreeType/contrib/ttf2pk/ttf2tfm.c
/*
 *   ttf2tfm.c
 *
 *   This file is part of the ttf2pk package.
 *
 *   Copyright 1997-1998 by
 *     Frederic Loyer <loyer@ensta.fr>
 *     Werner Lemberg <wl@gnu.org>.
 */

/*
 *   This program converts TTF files to TeX TFM files, and optionally
 *   to TeX VPL files that retain all kerning and ligature information.
 *   Both files make the characters not normally encoded by TeX available
 *   by character codes greater than 0x7F.
 */

/*
 *   Adapted from afm2tfm by F. Loyer <loyer@ensta.fr>.
 */

#include <stdio.h>
#include <stdlib.h>
#include <stddef.h>         /* for size_t */
#include <string.h>

#include "ttf2tfm.h"
#include "newobj.h"
#include "ttfenc.h"
#include "ligkern.h"
#include "texenc.h"
#include "ttfaux.h"
#include "tfmaux.h"
#include "vplaux.h"
#include "errormsg.h"
#include "filesrch.h"
#include "parse.h"
#include "subfont.h"


char ident[] = "ttf2tfm version 1.2";
char progname[] = "ttf2tfm";            /* for error/warning messages */

/* command line options */

static char    makevpl;                 /* can be 1 or 2 */
static Boolean pedantic;
static Boolean quiet;
static Boolean forceoctal;


/*
 *   Re-encode the TTF font.
 */

static void
handlereencoding(Font *fnt)
{
  int i;
  ttfinfo *ti;
  char *p;


  if (fnt->inencname)
  {
    fnt->inencoding = readencoding(&(fnt->inencname), fnt, True);

    /* reset all pointers in the mapping table */
    for (i = 0; i <= 0xFF; i++)
      if (NULL != (ti = fnt->inencptrs[i]))
      {
        ti->incode = -1;
        fnt->inencptrs[i] = NULL;
      }

    /*
     *   Reencode TTF <--> raw TeX.  Only these code points will be used
     *   for the output encoding.
     */

    for (i = 0; i <= 0xFF; i++)
    {
      p = fnt->inencoding->vec[i];
      if (p && *p)
        if ((ti = findadobe(p, fnt->charlist)))
        {
          if (ti->incode >= 0)
          {
            warning("Character `%s' encoded twice in input encoding\n"
                    "         (positions %x and %x; the latter is ignored).",
                    p, ti->incode, i);
            fnt->inencoding->vec[i] = ".notdef";
            continue;
          }
          if (ti->charcode >= 0)
          {
            ti->incode = i;
            fnt->inencptrs[i] = ti;
          }
        }
        else
        {
          warning("Cannot find character `%s'\n"
                  "         specified in input encoding.", p);
        }
    }
    fnt->codingscheme = fnt->inencoding->name;
  }

  if (!quiet)
  {
    if (fnt->inencname)
      printf("\nUsing %s as input encoding.\n", fnt->inencname);
    else
    {
      printf(
        "\nUsing the first 256 glyphs in the following input encoding:\n\n");
      for (i = 0; i <= 0xFF; i++)
      {
        if ((ti = fnt->inencptrs[i]))
          printf("  0x%02x    %s\n", i, ti->adobename);
      }
      printf("\n");
    }
  }

  if (fnt->outencname)
    fnt->outencoding = readencoding(&(fnt->outencname), fnt, False);
  else
    fnt->outencoding = readencoding(NULL, fnt, False);
}


static void
assignchars(Font *fnt)
{
  register char **p;
  register int i, j, k;
  register ttfinfo *ti;
  int nextfree = 0x80;


  /*
   *   First, we assign all those that match perfectly.
   */

  for (i = 0, p = fnt->outencoding->vec; i <= 0xFF; i++, p++)
    if ((ti = findmappedadobe(*p, fnt->inencptrs)))
    {
      if (ti->outcode >= 0)
        fnt->nextout[i] = ti->outcode;       /* linked list */
      ti->outcode = i;
      fnt->outencptrs[i] = ti;
    }
    else if (strcmp(*p, ".notdef") != 0)
      warning("Cannot map character `%s'\n"
              "         specified in output encoding.", *p);

  if (pedantic)
    goto end;

  /*
   *   Next, we assign all the others, retaining the TTF code positions,
   *   possibly multiplying assigned characters, unless the output encoding
   *   was precisely specified.
   */

  for (i = 0; i <= 0xFF; i++)
    if ((ti = fnt->inencptrs[i]) &&
         ti->charcode >= 0 && ti->charcode <= 0xFF &&
         ti->outcode < 0 && fnt->outencptrs[ti->charcode] == NULL)
    {
      ti->outcode = ti->charcode;
      fnt->outencptrs[ti->charcode] = ti;
    }

  /*
   *   Finally, we map all remaining characters into free locations beginning
   *   with 0x80.
   */

  for (i = 0; i <= 0xFF; i++)
    if ((ti = fnt->inencptrs[i]) && ti->outcode < 0)
    {
      while (fnt->outencptrs[nextfree])
      {
        nextfree = (nextfree + 1) & 0xFF;
        if (nextfree == 0x80)
          goto finishup;        /* all slots full */
      }
      ti->outcode = nextfree;
      fnt->outencptrs[nextfree] = ti;
    }

finishup:

  /*
   *   Now, if any of the characters are encoded multiple times, we want
   *   ti->outcode to be the first one assigned, since that is most likely
   *   to be the most important one.  So we reverse the above lists.
   */

  for (i = 0; i <= 0xFF; i++)
    if ((ti = fnt->inencptrs[i]) && ti->outcode >= 0)
    {
      k = -1;
      while (fnt->nextout[ti->outcode] >= 0)
      {
        j = fnt->nextout[ti->outcode];
        fnt->nextout[ti->outcode] = k;
        k = ti->outcode;
        ti->outcode = j;
      }
      fnt->nextout[ti->outcode] = k;
    }

end:

  if (!quiet)
  {
    printf("\nUsing the following output encoding:\n\n");
    for (i = 0; i <= 0xFF; i++)
    {
      if ((ti = fnt->outencptrs[i]))
        printf("  0x%02x    %s\n", i, ti->adobename);
    }
    printf("\n");
  }
}


#define VERSION "\
Copyright (C) 1997-1998 Frederic Loyer and Werner Lemberg.\n\
There is NO warranty.  You may redistribute this software\n\
under the terms of the GNU General Public License\n\
and the Dvips copyright.\n\
\n\
For more information about these matters, see the files\n\
named COPYING and ttf2tfm.c.\n\
\n\
Primary authors of ttf2tfm: F. Loyer and W. Lemberg.\n\
\n\
ttf2tfm is based on afm2tfm from T. Rokicki\n\
and the FreeType project from\n\
David Turner, Robert Wilhelm, and Werner Lemberg.\n\
"

static void
version(void)
{
  fputs(ident, stdout);
  fprintf(stdout, " (%s)\n", TeX_search_version());
  fputs(VERSION, stdout);
  exit(0);
}


#define USAGE "\
  Convert a TrueType font table to TeX's font metric format.\n\
\n\
-c REAL             use REAL for height of small caps made with -V [0.8]\n\
-e REAL             widen (extend) characters by a factor of REAL [1.0]\n\
-E INT              select INT as the TTF encoding ID [1]\n\
-f INT              select INT as the font index in a TTC [0]\n\
-n                  use PS names of TrueType font\n\
-N                  use only PS names and no cmap\n\
-O                  use octal for all character codes in the vpl file\n\
-p ENCFILE[.enc]    read ENCFILE for the TTF->raw TeX mapping\n\
-P INT              select INT as the TTF platform ID [3]\n\
-q                  suppress informational output\n\
-r OLDNAME NEWNAME  replace glyph name OLDNAME with NEWNAME\n\
-R RPLFILE[.rpl]    read RPLFILE containing glyph replacement names\n\
-s REAL             oblique (slant) characters by REAL, usually <<1 [0.0]\n\
-t ENCFILE[.enc]    read ENCFILE for the encoding of the vpl file\n\
-T ENCFILE[.enc]    equivalent to -p ENCFILE -t ENCFILE\n\
-u                  output only characters from encodings, nothing extra\n\
-v FILE[.vpl]       make a VPL file for conversion to VF\n\
-V SCFILE[.vpl]     like -v, but synthesize smallcaps as lowercase\n\
--help              print this message and exit\n\
--version           print version number and exit\n\
"

static void
usage(void)
{
  fputs("Usage: ttf2tfm FILE[.ttf|.ttc] [OPTION]... [FILE[.tfm]]\n", stdout);
  fputs(USAGE, stdout);
  exit(0);
}


static void
handle_options(int argc, char *argv[], Font *fnt)
{
  register int lastext;
  register int i;
  size_t l;
  int arginc;
  char *temp;
  char c;
  char *vpl_name = NULL;
  int have_capheight = 0;
  int have_sfd = 0;
  int sfd_begin, postfix_begin;
  int base_name;
  stringlist* sl;


  /* scan first whether the -q switch is set */
  for (i = 1; i < argc; i++)
    if (argv[i][0] == '-' && argv[i][1] == 'q')
      quiet = True;

  if (!quiet)
    printf("This is %s\n", ident);

#if defined(MSDOS) || defined(OS2) || defined(ATARIST)
  /* Make VPL file identical to that created under Unix */
  fnt->titlebuf = (char *)mymalloc(strlen(progname) + strlen(argv[1]) +
                                   1 + 1);
  sprintf(fnt->titlebuf, "%s %s", progname, argv[1]);
#else
  fnt->titlebuf = (char *)mymalloc(strlen(argv[0]) + strlen(argv[1]) +
                                   1 + 1);
  sprintf(fnt->titlebuf, "%s %s", argv[0], argv[1]);
#endif


  /*
   *   TrueType font name.
   */

  fnt->ttfname = newstring(argv[1]);

  /*
   *   The other arguments.  We delay the final processing of some switches
   *   until the tfm font name has been scanned -- if it contains two `@'s,
   *   many switches are ignored.
   */

  while (argc > 2 && *argv[2] == '-')
  {
    arginc = 2;
    i = argv[2][1];

    switch (i)
    {
    case 'v':
      makevpl = 1;
      if (argc <= 3)
        oops("Missing parameter for -v option.");
      if (vpl_name)
        free(vpl_name);
      vpl_name = newstring(argv[3]);
      handle_extension(&vpl_name, ".vpl");
      break;

    case 'V':
      makevpl = 2;
      if (argc <= 3)
        oops("Missing parameter for -V option.");
      if (vpl_name)
        free(vpl_name);
      vpl_name = newstring(argv[3]);
      handle_extension(&vpl_name, ".vpl");
      break;

    case 'f':
      if (argc <= 3)
        oops("Missing parameter for -f option.");
      if (sscanf(argv[3], "%lu", &(fnt->fontindex)) == 0)
        oops("Invalid font index.");
      fnt->fontindexparam = argv[3];
      break;

    case 'E':
      if (argc <= 3)
        oops("Missing parameter for -E option.");
      if (sscanf(argv[3], "%hu", &(fnt->eid)) == 0)
        oops("Invalid encoding ID.");
      fnt->eidparam = argv[3];
      break;

    case 'P':
      if (argc <= 3)
        oops("Missing parameter for -P option.");
      if (sscanf(argv[3], "%hu", &(fnt->pid)) == 0)
        oops("Invalid platform ID.");
      fnt->pidparam = argv[3];
      break;

    case 'e':
      if (argc <= 3)
        oops("Missing parameter for -e option.");
      if (sscanf(argv[3], "%f", &(fnt->efactor)) == 0 || fnt->efactor < 0.01)
        oops("Bad extension factor.");
      fnt->efactorparam = argv[3];
      break;

    case 'c':
      if (argc <= 3)
        oops("Missing parameter for -c option.");
      have_capheight = 1;
      if (sscanf(argv[3], "%f", &(fnt->capheight)) == 0)
        fnt->capheight = 0;
      break;

    case 's':
      if (argc <= 3)
        oops("Missing parameter for -s option.");
      if (sscanf(argv[3], "%f", &(fnt->slant)) == 0)
        oops("Bad slant parameter.");
      fnt->slantparam = argv[3];
      break;

    case 'p':
      if (argc <= 3)
        oops("Missing parameter for -p option.");
      if (fnt->inencname)
        free(fnt->inencname);
      fnt->inencname = newstring(argv[3]);
      break;

    case 'T':
      if (argc <= 3)
        oops("Missing parameter for -T option.");
      if (fnt->inencname)
        free(fnt->inencname);
      if (fnt->outencname)
        free(fnt->outencname);
      fnt->inencname = newstring(argv[3]);
      fnt->outencname = newstring(argv[3]);
      break;

    case 't':
      if (argc <= 3)
        oops("Missing parameter for -T option.");
      if (fnt->outencname)
        free(fnt->outencname);
      fnt->outencname = newstring(argv[3]);
      break;

    case 'r':
      if (argc <= 4)
        oops("Not enough parameters for -r option.");
      sl = newstringlist();
      sl->old_name = newstring(argv[3]);
      sl->new_name = newstring(argv[4]);
      sl->single_replacement = True;
      sl->next = fnt->replacements;
      fnt->replacements = sl;
      arginc = 3;
      break;

    case 'R':
      if (argc <= 3)
        oops("Missing parameter for -R option.");
      if (fnt->replacementname)
        free(fnt->replacementname);
      fnt->replacementname = newstring(argv[3]);
      break;

    case 'O':
      forceoctal = True;
      arginc = 1;
      break;

    case 'n':
      fnt->PSnames = Yes;
      arginc = 1;
      break;

    case 'N':
      fnt->PSnames = Only;
      arginc = 1;
      break;

    case 'u':
      pedantic = True;
      arginc = 1;
      break;

    case 'q':
      quiet = True;
      arginc = 1;
      break;

    default:
      if (argc <= 3 || argv[3][0] == '-')
      {
        warning("Unknown option `%s' will be ignored.\n", argv[2]);
        arginc = 1;
      }
      else
        warning("Unknown option `%s %s' will be ignored.\n",
                argv[2], argv[3]);
    }

    for (i = 0; i < arginc; i++)
    {
      l = strlen(fnt->titlebuf);
      fnt->titlebuf = (char *)myrealloc((void *)fnt->titlebuf,
                                        l + strlen(argv[2]) + 1 + 1);
      sprintf(fnt->titlebuf + strlen(fnt->titlebuf), " %s", argv[2]);
      argv++;
      argc--;
    }
  }

  /* Read replacement glyph name file */

  get_replacements(fnt);

  if (argc > 3 || (argc == 3 && *argv[2] == '-'))
    oops("Need at most two non-option arguments.");

  /*
   *   The tfm file name.
   */

  if (argc == 2)
    temp = newstring(fnt->ttfname);
  else
  {
    temp = newstring(argv[2]);
    l = strlen(fnt->titlebuf);
    fnt->titlebuf = (char *)myrealloc((void *)fnt->titlebuf,
                                      l + strlen(argv[2]) + 1 + 1);
    sprintf(fnt->titlebuf + strlen(fnt->titlebuf), " %s", argv[2]);
  }

  handle_sfd(temp, &sfd_begin, &postfix_begin);

  if (sfd_begin > -1)
  {
    have_sfd = True;
    i = sfd_begin - 2;
  }
  else
    i = strlen(temp) - 1;

  /*
   *   Now we search the beginning of the name without directory.
   */

  for (; i >= 0; i--)
    if (temp[i] == '/' || temp[i] == ':' || temp[i] == '\\')
      break;

  base_name = i + 1;

  /*
   *   We store the path (with the final directory separator).
   */

  if (base_name > 0)
  {
    c = temp[base_name];
    temp[base_name] = '\0';
    fnt->tfm_path = newstring(temp);
    temp[base_name] = c;
  }

  if (have_sfd)
  {
    /* the prefix and the sfd file name */

    if (temp[base_name])
      fnt->outname = newstring(temp + base_name);

    fnt->sfdname = newstring(temp + sfd_begin);
  }
  else
    postfix_begin = base_name;

  /*
   *   Get the extension.
   */

  lastext = -1;
  for (i = postfix_begin; temp[i]; i++)
    if (temp[i] == '.')
      lastext = i;

  if (argc == 2 && lastext >= 0)
  {
    temp[lastext] = '\0';       /* remove TTF file extension */
    lastext = -1;
  }

  if (lastext == -1)
    fnt->tfm_ext = newstring(".tfm");
  else
  {
    fnt->tfm_ext = newstring(temp + lastext);
    temp[lastext] = '\0';
  }

  if (have_sfd)
  {
    if (temp[postfix_begin])
      fnt->outname_postfix = newstring(temp + postfix_begin);
  }
  else
  {
    if (temp[base_name])
      fnt->outname = newstring(temp + base_name);
    else
      oops("Invalid tfm file name.");
  }


  /*
   *   Now we can process the remaining parameters.
   */

  if (have_sfd)
  {
    if (makevpl)
    {
      warning("Ignoring `-v' and `-V' switches for subfonts.");
      makevpl = 0;
    }
    if (have_capheight)
      warning("Ignoring `-c' switch for subfonts.");
    if (fnt->inencname || fnt->outencname)
    {
      warning("Ignoring `-p', `-t', and `-T' switches for subfonts.");
      fnt->inencname = NULL;
      fnt->outencname = NULL;
    }
    if (fnt->PSnames)
    {
      warning("Ignoring `-n' or '-N' switch for subfonts.");
      fnt->PSnames = No;
    }

    init_sfd(fnt, True);
  }
  else
  {
    if (have_capheight && fnt->capheight < 0.01)
      oops("Bad small caps height.");

    if (vpl_name)
      if ((fnt->vplout = fopen(vpl_name, "wt")) == NULL)
        oops("Cannot open vpl output file.");
  }

  if (fnt->PSnames == Only)
  {
    if (fnt->pidparam || fnt->eidparam)
    {
      warning("Ignoring `-P' and `-E' options if `-N' switch is selected.");
      fnt->pidparam = NULL;
      fnt->eidparam = NULL;
    }
  }

  if (vpl_name)
    free(vpl_name);
  free(temp);
}


/*
 *   This routine prints out the line that needs to be added to ttfonts.map.
 */

static void
consttfonts(Font *fnt)
{
  if (!quiet)
    printf("\n");
  printf("%s", fnt->outname);
  if (fnt->sfdname)
    printf("@%s@", fnt->sfdname);
  if (fnt->outname_postfix)
    printf("%s", fnt->outname_postfix);

  printf("   %s", fnt->ttfname);

  if (fnt->slantparam || fnt->efactorparam ||
      fnt->inencname ||
      fnt->pidparam || fnt->eidparam ||
      fnt->fontindexparam ||
      fnt->replacements ||
      fnt->replacementname ||
      fnt->PSnames)
  {
    if (fnt->slantparam)
      printf(" Slant=%s", fnt->slantparam);
    if (fnt->efactorparam)
      printf(" Extend=%s", fnt->efactorparam);
    if (fnt->inencname)
      printf(" Encoding=%s", fnt->inencname);
    if (fnt->pidparam)
      printf(" Pid=%s", fnt->pidparam);
    if (fnt->eidparam)
      printf(" Eid=%s", fnt->eidparam);
    if (fnt->fontindexparam)
      printf(" Fontindex=%s", fnt->fontindexparam);
    if (fnt->PSnames)
      printf(" PS=%s", fnt->PSnames == Yes ? "Yes" : "Only");
    if (fnt->replacementname && fnt->inencoding)
      printf(" Replacement=%s", fnt->replacementname);
    if (fnt->replacements && fnt->inencoding)
    {
      stringlist *sl;


      for (sl = fnt->replacements; sl; sl = sl->next)
        if (sl->single_replacement)
          printf(" %s=%s", sl->old_name, sl->new_name);
    }
  }
  printf("\n");
}


int
main(int argc, char *argv[])
{
  Font font;
  ttfinfo *ti;


  init_font_structure(&font);

  TeX_search_init(argv[0], "ttf2tfm", "TTF2TFM");

  if (argc == 1)
  {
    fputs("ttf2tfm: Need at least one file argument.\n", stderr);
    fputs("Try `ttf2tfm --help' for more information.\n", stderr);
    exit(1);
  }
  if (argc == 2)
  {
    if (strcmp(argv[1], "--help") == 0)
      usage();
    else if (strcmp(argv[1], "--version") == 0)
      version();
  }

  handle_options(argc, argv, &font);

  if (font.sfdname)
  {
    while (get_sfd(&font))
    {
      char *temp;
      int i, start, end, len;


      get_tfm_fullname(&font);

      /*
       *   Extract base name of sfd file.
       */

      temp = newstring(font.sfdname);
      len = strlen(temp);

      start = 0;
      for (i = len - 1; i >= 0; i--)
        if (temp[i] == '/' || temp[i] == ':' || temp[i] == '\\')
        {
          start = i + 1;
          break;
        }

      end = len;
      for (i = len - 1; i >= 0; i--)
        if (temp[i] == '.')
        {
          end = i;
          break;
        }
      temp[end] = '\0';

      font.codingscheme = (char *)mymalloc(strlen(temp + start) + 4 + 1);
      sprintf(font.codingscheme, "CJK-%s", temp + start);
      free(temp);

      readttf(&font, quiet, True);
      if (font.replacements)
        warning("Replacement glyphs will be ignored.");

      if (NULL != (ti = findadobe("space", font.charlist)))
        font.fontspace = ti->width;
      else if (NULL != (ti = findadobe(".c0x20", font.charlist)))
        font.fontspace = ti->width;
      else
        font.fontspace = transform(500, 0, font.efactor, font.slant);

      if (buildtfm(&font, True))
        writetfm(&font);
    }

    close_sfd();
  }
  else
  {
    get_tfm_fullname(&font);

    readttf(&font, quiet, False);
    replace_glyphs(&font);

    if (NULL != (ti = findadobe("space", font.charlist)))
      font.fontspace = ti->width;
    else if (NULL != (ti = findadobe(".c0x20", font.charlist)))
      font.fontspace = ti->width;
    else
      font.fontspace = transform(500, 0, font.efactor, font.slant);

    handlereencoding(&font);

    buildtfm(&font, False);
    writetfm(&font);
  }

  if (makevpl)
  {
    assignchars(&font);
    if (makevpl > 1)
      upmap(&font);
    writevpl(&font, makevpl, forceoctal);
    fclose(font.vplout);
  }

  consttfonts(&font);

  exit(0);      /* for safety reasons */
  return 0;     /* never reached */
}


/* end */
